### ISSUE TITLE
OnConnectionTypeCreated is not called when AvoidSpecificDataProviderAPI is true (PostgreSQL)

### ISSUE THREAD

**Issue by e673** [linq2db/linq2db#360]
I use a wrapper over `NpgsqlConnection` to connect to PostgreSQL database. In this case, the function `IsCompatibleConnection` returns false and I suppress it by either setting `LinqToDB.Common.Configuration.AvoidSpecificDataProviderAPI = true` or overriding in.

The problem is that `OnConnectionTypeCreated` is never called in this case: it is called only in `DynamicDataProviderBase.GetConnectionType()` which is called only from the function `DynamicDataProviderBase.IsCompatibleConnection()`.

So I have to initialize it manually by explicit calling `DataProvider.IsCompatibleConnection(...)` before performing any queries.

**Comment by igor-tkachev** [comment]
How can I reproduce it?

**Comment by e673** [comment]
```
    public class Schema : LinqToDB.Data.DataConnection
    {
        public Schema(IDbConnection dbcon)
            : base(new CustomPostgreSQLDataProvider(), dbcon) { }

        public ITable<UserAccount> UserAccounts { get { return GetTable<UserAccount>(); } }
    }

```
Where `dbcon` is NOT an instance of `Npgsql.NpgsqlConnection`, for example, a wrapper.

```
    class CustomPostgreSQLDataProvider
        : PostgreSQLDataProvider
    {
        public override bool IsCompatibleConnection(IDbConnection connection)
        {
            bool ignore = base.IsCompatibleConnection(connection);
            return true;
        }
    }
```

Then try to select anything containing DB-specific fields, for example, bool type.

**Comment by agoretsky** [comment]
same issue causes NullReferenceException

**Comment by agoretsky** [comment]
repro

```
using System.Data;
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.DataProvider;
using LinqToDB.DataProvider.PostgreSQL;
using Newtonsoft.Json;
using Npgsql;

namespace ConsoleApp35
{
    class Program
    {
        static void Main(string[] args)
        {
			using (var db = new DbDataContext())
	        {
				var item = new Entity
				{
					Data = JsonConvert.SerializeObject(new Data
					{
						X = 5,
						Y = 7
					})
				};

				var id = db.InsertWithInt64Identity(item);
	        }
		}
    }

	public class DbDataContext : DataConnection
	{
		public DbDataContext() : base(GetDataProvider(), GetConnection(), new MappingSchemaAccessor().MappingSchema )
		{
		}

		private static IDataProvider GetDataProvider()
		{
			LinqToDB.Common.Configuration.AvoidSpecificDataProviderAPI = true;
			return new PostgreSQLDataProvider(PostgreSQLVersion.v95);
		}

		private static IDbConnection GetConnection()
		{
			return new NpgsqlConnection(@"Host=127.0.0.1;Port=5432;Username=test;Password=test;Database=test");
		}

		public ITable<Entity> Entities => GetTable<Entity>();
	}

	public class MappingSchemaAccessor
	{
		/// <summary>
		/// Инициализация провайдера доступа.
		/// </summary>
		public MappingSchemaAccessor()
		{
			MappingSchema = LinqToDB.Mapping.MappingSchema.Default;
			MappingSchema.GetFluentMappingBuilder()
				.Entity<Entity>()
				.HasTableName("entity")
				.HasPrimaryKey(x => x.Id)
				.HasIdentity(x => x.Id)
				.Property(x => x.Id).HasColumnName("id")
				.Property(x => x.Data).HasColumnName("data").HasDataType(DataType.Json);
		}

		/// <summary>
		/// Схема маппинга.
		/// </summary>
		public LinqToDB.Mapping.MappingSchema MappingSchema { get; }
	}


	public class Entity
	{
		public long Id { get; set; }

		public string Data { get; set; }
	}

	public class Data
	{
		public long X { get; set; }

		public long Y { get; set; }
	}
}
```

**Comment by agoretsky** [comment]
docs example says about LinqToDB.Common.Configuration.AvoidSpecificDataProviderAPI = true

> // you can move this line to other place, but it should be allways set before LINQ to DB provider instance creation

I think there shoud be a note that it is only for custom DbConnection classes like ProfiledDbConnection

**Comment by agoretsky** [comment]
Either way, some kind of InvalidOperationException/Unsupported with details should be thrown instead of NullReference exception

**Comment by MaceWindu** [comment]
Duplicate of #1438

**Comment by MaceWindu** [comment]
closed by #1447
